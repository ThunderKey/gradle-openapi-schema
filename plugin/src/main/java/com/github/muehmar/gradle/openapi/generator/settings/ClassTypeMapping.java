package com.github.muehmar.gradle.openapi.generator.settings;

import com.github.muehmar.gradle.openapi.OpenApiSchemaGeneratorExtension;
import java.io.Serializable;
import java.util.Objects;

public class ClassTypeMapping implements Serializable {
  private final String fromClass;
  private final String toClass;
  private final String imports;

  public ClassTypeMapping(String fromClass, String toClass, String imports) {
    this.fromClass = fromClass;
    this.toClass = toClass;
    this.imports = imports;
  }

  public static ClassTypeMapping fromExtension(OpenApiSchemaGeneratorExtension.ClassMapping ext) {
    return new ClassTypeMapping(ext.getFromClass(), ext.getToClass(), ext.getImports());
  }

  public String getFromClass() {
    return fromClass;
  }

  public String getToClass() {
    return toClass;
  }

  public String getImports() {
    return imports;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    ClassTypeMapping that = (ClassTypeMapping) o;
    return Objects.equals(fromClass, that.fromClass)
        && Objects.equals(toClass, that.toClass)
        && Objects.equals(imports, that.imports);
  }

  @Override
  public int hashCode() {
    return Objects.hash(fromClass, toClass, imports);
  }

  @Override
  public String toString() {
    return "ClassTypeMapping{"
        + "fromClass='"
        + fromClass
        + '\''
        + ", toClass='"
        + toClass
        + '\''
        + ", imports='"
        + imports
        + '\''
        + '}';
  }
}
