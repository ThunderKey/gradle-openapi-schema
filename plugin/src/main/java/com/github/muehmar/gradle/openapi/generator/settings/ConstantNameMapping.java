package com.github.muehmar.gradle.openapi.generator.settings;

import java.io.Serializable;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class ConstantNameMapping implements Serializable {
  private final String constant;
  private final String replacement;

  public ConstantNameMapping(String constant, String replacement) {
    this.constant = constant;
    this.replacement = replacement;
  }

  public String getConstant() {
    return constant;
  }

  public String getReplacement() {
    return replacement;
  }
}
