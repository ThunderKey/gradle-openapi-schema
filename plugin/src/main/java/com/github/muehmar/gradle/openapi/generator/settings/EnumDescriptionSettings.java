package com.github.muehmar.gradle.openapi.generator.settings;

import com.github.muehmar.gradle.openapi.generator.data.Name;
import java.io.Serializable;
import java.util.Objects;

/** Settings how and if the description of enum should get extracted. */
public class EnumDescriptionSettings implements Serializable {
  public static final String ENUM_PLACEHOLDER = "__ENUM__";
  private final boolean enabled;
  private final String prefixMatcher;
  private final boolean failOnIncompleteDescriptions;

  public EnumDescriptionSettings(
      boolean enabled, String prefixMatcher, boolean failOnIncompleteDescriptions) {
    if (enabled && !prefixMatcher.contains(ENUM_PLACEHOLDER)) {
      throw new IllegalArgumentException(
          "Prefix must contain the enum placeholder '"
              + ENUM_PLACEHOLDER
              + "' to properly match the correct description. But was '"
              + prefixMatcher
              + "'");
    }

    this.enabled = enabled;
    this.prefixMatcher = prefixMatcher;
    this.failOnIncompleteDescriptions = failOnIncompleteDescriptions;
  }

  public static EnumDescriptionSettings disabled() {
    return new EnumDescriptionSettings(false, "", false);
  }

  public static EnumDescriptionSettings enabled(
      String prefixMatcher, boolean failOnIncompleteDescriptions) {
    return new EnumDescriptionSettings(true, prefixMatcher, failOnIncompleteDescriptions);
  }

  public boolean isEnabled() {
    return enabled;
  }

  public boolean isDisabled() {
    return !isEnabled();
  }

  public String getPrefixMatcher() {
    return prefixMatcher;
  }

  public String getPrefixMatcherForMember(Name member) {
    return getPrefixMatcher().replace(ENUM_PLACEHOLDER, member.asString());
  }

  public boolean isFailOnIncompleteDescriptions() {
    return failOnIncompleteDescriptions;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    EnumDescriptionSettings that = (EnumDescriptionSettings) o;
    return enabled == that.enabled
        && failOnIncompleteDescriptions == that.failOnIncompleteDescriptions
        && Objects.equals(prefixMatcher, that.prefixMatcher);
  }

  @Override
  public int hashCode() {
    return Objects.hash(enabled, prefixMatcher, failOnIncompleteDescriptions);
  }

  @Override
  public String toString() {
    return "EnumDescriptionSettings{"
        + "enabled="
        + enabled
        + ", prefixMatcher='"
        + prefixMatcher
        + '\''
        + ", failOnIncompleteDescriptions="
        + failOnIncompleteDescriptions
        + '}';
  }
}
