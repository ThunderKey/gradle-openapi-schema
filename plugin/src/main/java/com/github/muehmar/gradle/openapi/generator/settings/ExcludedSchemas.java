package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.PojoSchema;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import java.util.function.Predicate;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class ExcludedSchemas {
  private final PList<Name> excludedNames;

  private ExcludedSchemas(PList<Name> excludedNames) {
    this.excludedNames = excludedNames;
  }

  public static ExcludedSchemas fromExcludedPojoNames(PList<Name> excludedNames) {
    return new ExcludedSchemas(excludedNames);
  }

  public Predicate<PojoSchema> getSchemaFilter() {
    return pojoSchema -> {
      final Predicate<Name> matchesPojoName =
          name -> pojoSchema.getPojoName().getName().equalsIgnoreCase(name);
      final Predicate<Name> matchesSchemaName =
          name -> pojoSchema.getSchemaName().asName().equalsIgnoreCase(name);

      return not(
          excludedNames.exists(
              excludedName ->
                  matchesPojoName.test(excludedName) || matchesSchemaName.test(excludedName)));
    };
  }
}
