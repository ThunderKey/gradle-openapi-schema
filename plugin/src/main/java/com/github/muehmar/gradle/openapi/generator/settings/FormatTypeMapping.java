package com.github.muehmar.gradle.openapi.generator.settings;

import java.io.Serializable;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class FormatTypeMapping implements Serializable {
  private final String formatType;
  private final String classType;

  public FormatTypeMapping(String formatType, String classType) {
    this.formatType = formatType;
    this.classType = classType;
  }

  public String getFormatType() {
    return formatType;
  }

  public String getClassType() {
    return classType;
  }
}
