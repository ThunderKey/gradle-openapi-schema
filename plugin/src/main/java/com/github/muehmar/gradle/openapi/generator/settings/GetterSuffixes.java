package com.github.muehmar.gradle.openapi.generator.settings;

import io.github.muehmar.pojoextension.annotations.SafeBuilder;
import java.io.Serializable;
import lombok.Value;

@Value
@SafeBuilder
public class GetterSuffixes implements Serializable {
  String requiredSuffix;
  String requiredNullableSuffix;
  String optionalSuffix;
  String optionalNullableSuffix;
}
