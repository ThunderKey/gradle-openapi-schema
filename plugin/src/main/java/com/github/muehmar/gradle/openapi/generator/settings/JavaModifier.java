package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import java.io.Serializable;
import java.util.Optional;

public enum JavaModifier implements Serializable {
  PRIVATE("private"),
  PROTECTED("protected"),
  PACKAGE_PRIVATE("package-private"),
  PUBLIC("public");

  private final String value;

  JavaModifier(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  public static Optional<JavaModifier> fromString(String value) {
    return PList.of(values()).find(modifier -> modifier.value.equalsIgnoreCase(value));
  }
}
