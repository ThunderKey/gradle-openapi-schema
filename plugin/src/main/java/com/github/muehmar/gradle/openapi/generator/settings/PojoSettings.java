package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
