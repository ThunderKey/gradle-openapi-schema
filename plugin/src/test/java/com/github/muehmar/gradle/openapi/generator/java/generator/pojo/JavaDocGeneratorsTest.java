package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.junit.jupiter.api.Assertions.*;

import au.com.origin.snapshots.Expect;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class JavaDocGeneratorsTest {

  private Expect expect;

  @Test
  void generate_when_disabledDeprecationWarning_then_noOutput() {
    final Generator<String, PojoSettings> gen =
        JavaDocGenerators.deprecatedValidationMethodJavaDoc();

    final Writer writer = gen.generate("", defaultTestSettings(), javaWriter());

    assertEquals("", writer.asString());
  }

  @Test
  void generate_when_enabledDeprecationWarning_then_correctOutput() {
    final Generator<String, PojoSettings> gen =
        JavaDocGenerators.deprecatedValidationMethodJavaDoc();

    final Writer writer =
        gen.generate(
            "",
            defaultTestSettings()
                .withValidationMethods(
                    TestPojoSettings.defaultValidationMethods().withDeprecatedAnnotation(true)),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
