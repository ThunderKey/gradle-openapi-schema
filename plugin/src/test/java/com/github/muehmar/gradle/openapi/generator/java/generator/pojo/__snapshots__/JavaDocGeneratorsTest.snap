com.github.muehmar.gradle.openapi.generator.java.generator.pojo.JavaDocGeneratorsTest.generate_when_enabledDeprecationWarning_then_correctOutput=[
.
.
/**
 * @deprecated This method is intended to be used only by frameworks for
 * validation or serialisation.
 */
]